core/math/qcp.cpp

```diff
@@ -0,0 +1,261 @@
+		} else {
+			double q0 = Math::sqrt(0.5 * (1.0 + dot / norm_product));
+			double coeff = 1.0 / (2.0 * q0 * norm_product);
+			Vector3 q = v.cross(u).normalized();
+			result = Quaternion(coeff * q.x, coeff * q.y, coeff * q.z, q0).normalized();
+		}
+	} else {
+		double a13 = -sum_xz_minus_zx;
+		double a14 = sum_xy_minus_yx;
+		double a21 = sum_yz_minus_zy;
+		double a22 = sum_xx_minus_yy - sum_zz - max_eigenvalue;
+		double a23 = sum_xy_plus_yx;
+		double a24 = sum_xz_plus_zx;
+		double a31 = a13;
+		double a32 = a23;
+		double a33 = sum_yy - sum_xx - sum_zz - max_eigenvalue;
+		double a34 = sum_yz_plus_zy;
+		double a41 = a14;
+		double a42 = a24;
+		double a43 = a34;
+		double a44 = sum_zz - sum_xx_plus_yy - max_eigenvalue;
+
+		double a3344_4334 = a33 * a44 - a43 * a34;
+		double a3244_4234 = a32 * a44 - a42 * a34;
+		double a3243_4233 = a32 * a43 - a42 * a33;
+		double a3143_4133 = a31 * a43 - a41 * a33;
+		double a3144_4134 = a31 * a44 - a41 * a34;
+		double a3142_4132 = a31 * a42 - a41 * a32;
+
+		double quaternion_w = a22 * a3344_4334 - a23 * a3244_4234 + a24 * a3243_4233;
+		double quaternion_x = -a21 * a3344_4334 + a23 * a3144_4134 - a24 * a3143_4133;
+		double quaternion_y = a21 * a3244_4234 - a22 * a3144_4134 + a24 * a3142_4132;
+		double quaternion_z = -a21 * a3243_4233 + a22 * a3143_4133 - a23 * a3142_4132;
+		double qsqr = quaternion_w * quaternion_w + quaternion_x * quaternion_x + quaternion_y * quaternion_y + quaternion_z * quaternion_z;
+
+		if (qsqr < eigenvector_precision) {
+			result = Quaternion();
+		} else {
+			quaternion_x *= -1;
+			quaternion_y *= -1;
+			quaternion_z *= -1;
+			double min = quaternion_w;
+			min = quaternion_x < min ? quaternion_x : min;
+			min = quaternion_y < min ? quaternion_y : min;
+			min = quaternion_z < min ? quaternion_z : min;
+			quaternion_w /= min;
+			quaternion_x /= min;
+			quaternion_y /= min;
+			quaternion_z /= min;
+			result = Quaternion(quaternion_x, quaternion_y, quaternion_z, quaternion_w).normalized();
+		}
+	}
+
+	return result;
+}
+
+void QuaternionCharacteristicPolynomial::translate(Vector3 r_translate, PackedVector3Array &r_x) {
+	for (Vector3 &p : r_x) {
+		p += r_translate;
+	}
+}
+
+Vector3 QuaternionCharacteristicPolynomial::_get_translation() {
+	return target_center - moved_center;
+}
+
+Vector3 QuaternionCharacteristicPolynomial::move_to_weighted_center(PackedVector3Array &r_to_center, Vector<double> &r_weight) {
+	Vector3 center;
+	double total_weight = 0;
+	bool weight_is_empty = r_weight.is_empty();
+	int size = r_to_center.size();
+
+	for (int i = 0; i < size; i++) {
+		if (!weight_is_empty) {
+			total_weight += r_weight[i];
+			center += r_to_center[i] * r_weight[i];
+		} else {
+			center += r_to_center[i];
+			total_weight++;
+		}
+	}
+
+	if (total_weight > 0) {
+		center /= total_weight;
+	}
+
+	return center;
+}
+
+void QuaternionCharacteristicPolynomial::inner_product(PackedVector3Array &coords1, PackedVector3Array &coords2) {
```

Comment:
Godot Engine uses `r_` and `p_` for passed through and update through the parameter.